File: src/main/java/tech/flandia_yingm/auto_fgo/img/Template.java
package tech.flandia_yingm.auto_fgo.img;

import lombok.NonNull;
import lombok.ToString;
import lombok.Value;

import java.awt.image.BufferedImage;

@Value
@ToString(onlyExplicitlyIncluded = true)
public class Template {

    @NonNull
    @ToString.Include
    private final String name;

    @NonNull
    private final BufferedImage image;

    private double threshold;


    public static Template of(String name, BufferedImage image, double threshold) {
        return new Template(name, image, threshold);
    }

}

File: src/main/java/tech/flandia_yingm/auto_fgo/img/Template.kt
package tech.flandia_yingm.auto_fgo.img

import tech.flandia_yingm.auto_fgo.img.Images.readResource
import java.awt.image.BufferedImage
import java.awt.image.BufferedImage.TYPE_INT_ARGB

data class Template(val name: String, val image: BufferedImage, val threshold: Double = 0.95) {

    companion object {
        val EMPTY_TEMPLATE = Template("", BufferedImage(1, 1, TYPE_INT_ARGB))
    }

    override fun toString(): String = name
}

fun template(name: String, threshold: Double = 0.95, clazz: () -> Unit): Template {
    return Template(name, readResource("$name.png", Class.forName(name(clazz))), threshold)
}

internal fun name(func: () -> Unit): String {
    val name = func.javaClass.name
    return when {
        name.contains("Kt$") -> name.substringBefore("Kt$")
        name.contains("$") -> name.substringBefore("$")
        else -> name
    }
}